airbyte-oauth/src/main/java/io/airbyte/oauth/flows/GitlabOAuthFlow.java

```diff
@@ -0,0 +1,111 @@
+    super(configRepository, httpClient, stateSupplier);
+  }
+
+  protected static String getDomain(JsonNode inputOAuthConfiguration) throws IOException {
+    final var domain = inputOAuthConfiguration.get("domain");
+    if (domain == null) {
+      throw new IOException("Domain field is empty.");
+    }
+    return domain.asText();
+  }
+
+  @Override
+  protected String formatConsentUrl(final UUID definitionId, final String clientId, final String redirectUrl, final JsonNode inputOAuthConfiguration)
+      throws IOException {
+    final URIBuilder builder = new URIBuilder()
+        .setScheme("https")
+        .setHost(getDomain(inputOAuthConfiguration))
+        .setPath("oauth/authorize")
+        .addParameter("client_id", clientId)
+        .addParameter("redirect_uri", redirectUrl)
+        .addParameter("state", getState())
+        .addParameter("response_type", "code")
+        .addParameter("scope", "read_api");
+    try {
+      return builder.build().toString();
+    } catch (URISyntaxException e) {
+      throw new IOException("Failed to format Consent URL for OAuth flow", e);
+    }
+  }
+
+  @Override
+  protected String getAccessTokenUrl(final JsonNode inputOAuthConfiguration) {
+    final var domain = inputOAuthConfiguration.get("domain");
+    return String.format(ACCESS_TOKEN_URL, domain == null ? "gitlab.com" : domain.asText());
+  }
+
+  @Override
+  protected Map<String, String> getAccessTokenQueryParameters(final String clientId,
+                                                              final String clientSecret,
+                                                              final String authCode,
+                                                              final String redirectUrl) {
+    return ImmutableMap.<String, String>builder()
+        .put("client_id", clientId)
+        .put("client_secret", clientSecret)
+        .put("code", authCode)
+        .put("grant_type", "authorization_code")
+        .put("redirect_uri", redirectUrl)
+        .build();
+  }
+
+  @Override
+  protected Map<String, Object> extractOAuthOutput(final JsonNode data, final String accessTokenUrl) throws IOException {
+    final Map<String, Object> result = new HashMap<>();
+    // check for refresh_token after successful authentication
+    if (data.has("refresh_token")) {
+      result.put("refresh_token", data.get("refresh_token").asText());
+    } else {
+      throw new IOException(String.format("Missing 'refresh_token' in query params from %s", accessTokenUrl));
+    }
+    // check for access_token after successful authentication
+    if (data.has("access_token")) {
+      result.put("access_token", data.get("access_token").asText());
+    } else {
+      throw new IOException(String.format("Missing 'access_token' in query params from %s", accessTokenUrl));
+    }
+    return result;
+  }
+
+  @Override
+  @Deprecated
+  public Map<String, Object> completeSourceOAuth(final UUID workspaceId,
```

Comment:
should we add this as a default implementation in the Interface instead of requiring all new flows to implement a deprecated method?

Reply:
this specific line does not require a reset, but the changes to the spec file do, yes

Reply:
@girarda afaik java interfaces do not support any kind of implementation - just signatures. If you mean base class `BaseOAuth2Flow` - there are already two implementations of `completeSourceOAuth` with different signatures, one of which is deprecated. If we change it, I'm afraid it would break some of the existing oauth flow implementations.
The trick here is that deprecated APIs can not support oauth flow Implementations with dependencies on connector config (that's exactly what Gitlab implementation is) so this method is overridden here and throws an exception. Please let me know if I'm missing something.

Reply:
we can use the [default keyword](https://stackoverflow.com/questions/31578427/what-is-the-purpose-of-the-default-keyword-in-java) to provide a default implementation.

It would look something like this
```
    default Map<String, Object> completeSourceOAuth(UUID workspaceId,
                                                    UUID sourceDefinitionId,
                                                    Map<String, Object> queryParams,
                                                    String redirectUrl,
                                                    JsonNode inputOAuthConfiguration,
                                                    OAuthConfigSpecification oauthConfigSpecification)
            throws IOException, ConfigNotFoundException, JsonValidationException {
        throw new IOException("Deprecated API. Do not use.");
    }
```

I completely agree that we shouldn't modify existing implementations. Too big of a blast radius for this PR

Reply:
oh thanks for that! nevertheless it looks like we wont extract much benefit from it. This method is anyway implemented in a `BaseOAuth2Flow` so will override the default implementation, and all of the existing flow implementations (and very likely future ones) subclass `BaseOAuth2Flow`. If you insist, I will add the default implementation but in the Gitlab flow we'll have to declare it as well

Reply:
> so will override the default implementation, and all of the existing flow implementations
fair enough. let's not do that then